### DISCUSSION TITLE
_grad_test variable type using ADKernel? I'm getting an error saying multiplication is an invalid operation.

### DISCUSSION THREAD

**Discussion by richmondodufisan** [idaholab/moose#18146] [Q&A General] [answered]
Hi, I'm completely new to Moose. I was following the tutorial posted to YouTube and modifying things here and there, but I'm having trouble computing the residual for a simple diffusion equation. What I have is very similar to the DarcyPressure.C tutorial, but the error message I get says the multiplication "*" operand is invalid. Below is the very simple kernel I'm trying to implement:

`
#include "FlowNet.h"
#include "DerivativeMaterialInterface.h"

registerMooseObject("farmApp", FlowNet);

InputParameters
FlowNet::validParams()
{
  InputParameters params = ADKernel::validParams();
  params.addClassDescription("Calculate flow potential under flownet");

  return params;
}

FlowNet::FlowNet(const InputParameters & parameters)
  : DerivativeMaterialInterface<ADKernel>(parameters),

    _permeability(getMaterialPropertyByName<Real>("permeability"))
{
}

ADReal 
FlowNet::computeQpResidual()
{
	/// isotropic permeability
	const auto & k = _permeability;

	return k * _grad_test[_i][_qp] * _grad_u[_qp];
}
`

with the error message:

error: no match for 'operator*' (operand types are 'const MaterialProperty<double>' and 'const value_type' {aka 'const libMesh::VectorValue<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<50> >, true> >'})

I am also aware that if I use the "ADKernelGrad" then I won't need to explicitly add "_grad_test" myself, but I wanted to do it with ADKernel.

**Comment by GiudGiud** [comment, marked as answer]
k is a material property right?
So it'd be k[_qp] not k to get the value at the quadrature point

**Reply by richmondodufisan** [reply]
Yes, its a scalar value for the permeability. Would it still be k[_qp] in that case? I don't see how, since it is a single value. I tried it though, and it worked! But I'm not sure why.

The example on the tutorial also left it as is (didn't index the quadrature point). (https://youtu.be/2tJwBsYaLaI @3:49:19) mark.

**Reply by GiudGiud** [reply]
Well in the header for that source file, you can see if it s a scalar (Real or ADReal) or a MaterialProperty

You may set it as a scalar in the input file, but MOOSE can convert scalars to material properties when needed

**Reply by richmondodufisan** [reply]
I see. That makes sense. Thank you!
